src/lem/constrainer.rs

```diff
@@ -81,145 +116,390 @@ impl LEM {
+        if Self::on_concrete_path(concrete_path)? {
+            let Some(ptr) = valuation.ptrs.get(name) else {
+                bail!("Couldn't retrieve {} from valuation", name);
+            };
+            store.hydrate_ptr(ptr)
+        } else {
+            Ok(ZPtr::dummy())
+        }
+    }
+
+    /// Accumulates slot data that will be used later to generate the constraints.
+    fn acc_slots_data_data<F: LurkField>(
+        concrete_path: Boolean,
+        slots_data: &mut Vec<SlotData<F>>,
+        img: AllocatedNum<F>,
+        preimg_vec: Vec<&AllocatedPtr<F>>,
+    ) {
+        let arity = match preimg_vec.len() {
+            2 => HashArity::A2,
+            3 => HashArity::A3,
+            4 => HashArity::A4,
+            _ => unreachable!(),
+        };
+        let mut preimg = Vec::with_capacity(preimg_vec.len() * 2);
+        for elem in preimg_vec {
+            preimg.push(elem.tag().clone());
+            preimg.push(elem.hash().clone());
+        }
+        slots_data.push(SlotData {
+            arity,
+            concrete_path,
+            preimg,
+            img,
+        });
+    }
+
+    /// Use the implies logic to constrain tag and hash values for accumulated
+    /// slot data
+    fn create_slot_constraints<F: LurkField, CS: ConstraintSystem<F>>(
+        cs: &mut CS,
+        slots_data: &Vec<SlotData<F>>,
+        store: &mut Store<F>,
+        alloc_manager: &mut AllocationManager<F>,
+        num_hash_slots: &NumSlots,
+    ) -> Result<()> {
+        let alloc_dummy_hash = alloc_manager.get_or_alloc_any(cs)?;
+
+        let mut hash2_count = 0;
+        let mut hash3_count = 0;
+        let mut hash4_count = 0;
+
+        macro_rules! constrain_slot {
+            (
+                $concrete_path: expr,
+                $hash_index: expr,
+                $preimg: expr,
+                $img: expr,
+                $constants: expr
+            ) => {
+                let alloc_hash = hash_poseidon(
+                    &mut cs.namespace(|| format!("slot_hash{}_{}", $preimg.len() / 2, $hash_index)),
+                    $preimg.to_vec(),
+                    $constants,
+                )?;
+
+                implies_equal(
+                    &mut cs.namespace(|| {
+                        format!(
+                            "implies equal hash for hash{}_{}",
+                            $preimg.len() / 2,
+                            $hash_index
+                        )
+                    }),
+                    $concrete_path,
+                    $img,
+                    &alloc_hash,
+                )?;
+            };
+        }
+
+        for SlotData {
+            arity,
+            concrete_path,
+            preimg,
+            img,
+        } in slots_data
+        {
+            // First we constrain concrete path hashes, later we
+            // fill with dummies in order to always use the same number of hashes.
+            if Self::on_concrete_path(concrete_path)? {
```

Comment:
This scares me, because it is either creating constraints ('constraining') or not, based on the *value* in a given instance/reduction/witness of a `Boolean`. But in order to have a uniform circuit, we can't do that. Therefore, in order for this to be correct, there must be some compensatory conditionals later, which manage to fill in the missing `else` clause of this `if`. I feel strongly that it would be better to solve this problem by including the `else` clause here and ensuring the same constraints are created. But this phrasing makes clear that it's probably better not even to have the if, and instead to just create the constraints once with the correct value.

In reality, the `Boolean` is supposed to hold all the information needed and shouldn't need to be 'examined' at all. The constraints themselves should do the work of handling either the true or false case.

Reply:
This `if` is gone thanks to @emmorais 🎉